### PR TITLE
Mint short otoken functionality

### PR DESCRIPTION
**Pull request by haythem96**
# Task: Feature Name:

Controller: mint short otoken

## High Level Description

This PR add the needed code for the 
- mint short otoken functionality
- check if collateral asset used and short otoken collateral `_isMarginableCollateral()`

Should be reviewed/merged after #109 

### Code

- [ ] Unit test 100% coverage
- [x] Does your code follow the naming and code documentation guidelines?

### Documentation

- [x] Is your code up to date with the spec? 
- [x] Have you added your tests to the testing doc?

### CODE REVIEW

File: contracts/MarginCalculator.sol [hunk 1]
```diff
@@ -59,6 +59,8 @@ contract MarginCalculator is Initializable {
     function getExcessCollateral(MarginAccount.Vault memory _vault) public view returns (uint256, bool) {
         // ensure the number of collateral, long and short array is valid.
         _checkIsValidSpread(_vault);
+        // ensure that the collateral asset is valid for the short asset
+        require(_isMarginableCollateral(_vault), "MarginCalculator: collateral asset nor marginable for short asset");
```
Comment (antoncoding) [inline contracts/MarginCalculator.sol:63]:
```suggestion
        require(_isMarginableCollateral(_vault), "MarginCalculator: collateral asset not marginable for short asset");
```

File: contracts/Controller.sol [hunk 2]
```diff
@@ -454,7 +464,26 @@ contract Controller is ReentrancyGuard, Ownable {
      * @dev only account owner or operator can withdraw long option from vault
      * @param _args MintArgs structure
      */
-    // function _mintOtoken(Actions.MintArgs memory _args) internal isAuthorized(_args.owner) {}
+    function _mintOtoken(Actions.MintArgs memory _args)
+        internal
+        isAuthorized(msg.sender, _args.owner)
```
Comment (antoncoding) [inline contracts/Controller.sol:469]: we can remove the `msg.sender` and just access it directly in the `isAuthorized` modifier right?

File: test/controller.test.ts [hunk 3]
```diff
@@ -1502,6 +1502,398 @@ contract('Controller', ([owner, accountOwner1, accountOperator1, random]) => {
...
+
+      longOtoken = await MockOtoken.new()
+      shortOtoken = await MockOtoken.new()
+      // init otoken
+      await longOtoken.init(
+        addressBook.address,
+        weth.address,
+        usdc.address,
+        usdc.address,
+        new BigNumber(250).times(new BigNumber(10).exponentiatedBy(18)),
```
Comment (antoncoding) [inline test/controller.test.ts:1520]: I have a function called `createScaledNumber()` from `utils.ts`, you can use that for creating BigNumber scaled by 1e18.

File: test/controller.test.ts [hunk 4]
```diff
@@ -1502,6 +1502,398 @@ contract('Controller', ([owner, accountOwner1, accountOperator1, random]) => {
...
+          {
+            actionType: ActionType.DepositCollateral,
+            owner: accountOwner1,
+            sender: random,
+            asset: usdc.address,
+            vaultId: vaultCounter.toNumber(),
+            amount: collateralToDeposit.toNumber(),
+            index: '0',
+            data: ZERO_ADDR,
+          },
```
Comment (antoncoding) [inline test/controller.test.ts:1569]: We can remove this action? (It looks like the test should still fail)

File: test/controller.test.ts [hunk 5]
```diff
@@ -1502,6 +1502,398 @@ contract('Controller', ([owner, accountOwner1, accountOperator1, random]) => {
...
+        ]
+
+        await expectRevert(
+          controller.operate(actionArgs, {from: accountOwner1}),
+          'Controller: invalid final vault state',
+        )
+      })
+
+      it('should withdraw exceeded collateral from naked short position when net value ', async () => {
```
Comment (antoncoding) [inline test/controller.test.ts:1777]:
```suggestion
      it('should withdraw exceeded collateral from naked short position when net value > 0 ', async () => {
```

**Review by antoncoding — APPROVED** [review]
Looks good

File: test/controller.test.ts (hunk 4 above)
Comment (haythem96) [reply test/controller.test.ts:1569]: Yeah I don't know what that action is doing there 👍

File: contracts/Controller.sol (hunk 2 above)
Comment (haythem96) [reply contracts/Controller.sol:469]: Yeah, I thought that can give more idea about it, who is authorized to access what, from the modifier args...

File: test/controller.test.ts (hunk 3 above)
Comment (haythem96) [reply test/controller.test.ts:1520]: Yeah, but that function return string, I like using BigNumber everywhere, convert to string only at `assert`.

File: contracts/Controller.sol (hunk 2 above)
Comment (antoncoding) [reply contracts/Controller.sol:469]: Good point

File: contracts/Controller.sol [hunk 6]
```diff
@@ -454,7 +464,26 @@ contract Controller is ReentrancyGuard, Ownable {
      * @dev only account owner or operator can withdraw long option from vault
```
Comment (aparnakr) [inline contracts/Controller.sol:464]:
```suggestion
     * @dev only account owner or operator can mint option from vault
```

File: contracts/Controller.sol [hunk 7]
```diff
@@ -454,7 +464,26 @@ contract Controller is ReentrancyGuard, Ownable {
...
      */
-    // function _mintOtoken(Actions.MintArgs memory _args) internal isAuthorized(_args.owner) {}
+    function _mintOtoken(Actions.MintArgs memory _args)
+        internal
+        isAuthorized(msg.sender, _args.owner)
+        returns (MarginAccount.Vault memory)
+    {
+        require(checkVaultId(_args.owner, _args.vaultId), "Controller: invalid vault id");
+        require(_args.to == msg.sender, "Controller: minter address and msg.sender address mismatch");
+
```
Comment (aparnakr) [inline contracts/Controller.sol:474]: does the check on isValidOtoken happen in margin calculator? (Like we shouldn't allow someone to mint a non Opyn otoken right?)

Comment (aparnakr) [reply contracts/Controller.sol:474]: Yeah, I don't think calculator checks this, so need that check here. isWhitelistedOtoken(short)

Comment (haythem96) [reply contracts/Controller.sol:474]: Yeah.

File: test/controller.test.ts [hunk 8]
```diff
@@ -1502,6 +1502,425 @@ contract('Controller', ([owner, accountOwner1, accountOperator1, random]) => {
...
+
+        await usdc.approve(marginPool.address, collateralToDeposit, {from: accountOwner1})
+        await expectRevert(
+          controller.operate(actionArgs, {from: accountOwner1}),
+          'MarginCalculator: Too many short otokens in the vault.',
+        )
+      })
+    })
+  })
+
```
Comment (aparnakr) [inline test/controller.test.ts:1923]: test: should only be able to mint whitelisted short tokens

**Review by aparnakr — COMMENTED** [review]
Looks good, needs to check that only valid short tokens can be minted.

File: contracts/Controller.sol (hunk 7 above)
Comment (haythem96) [reply contracts/Controller.sol:474]: Will add it.

**Comment by haythem96** [comment]
> Looks good, needs to check that only valid short tokens can be minted.

Added that.

**Review by aparnakr — APPROVED** [review]

**Pull request merged**